src/parse/parse-token-request.ts

```diff
@@ -0,0 +1,299 @@
+  };
+};
+
+const parseClientAssertion = (
+  clientAssertion: string,
+  tokenRequestClientId?: string
+): PrivateKeyJwt => {
+  const jwtParts = clientAssertion.split(".");
+
+  if (jwtParts.length !== 3) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Unexpected number of Base64URL parts, must be three"
+    );
+  }
+
+  const payload = decodeJwt(clientAssertion);
+  const header = decodeProtectedHeader(clientAssertion);
+  const signature = jwtParts[2];
+
+  if (!signature || signature.trim().length === 0) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: The signature must not be empty"
+    );
+  }
+
+  if (!payload.sub) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing subject in client JWT assertion"
+    );
+  }
+
+  if (!payload.iss) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing issuer in client JWT assertion"
+    );
+  }
+
+  if (!payload.aud) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing audience in client JWT assertion"
+    );
+  }
+
+  if (!payload.exp || typeof payload.exp !== "number") {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing or invalid expiry in client JWT assertion"
+    );
+  }
+
+  if (payload.nbf) {
+    //We don't care about this claim but it can be included and
+    //can cause a ParseError if the claim is invalid
+    logger.warn(
+      "nbf claim included in client JWT assertion, this claim is not needed: https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/authenticate-your-user/#create-a-jwt"
+    );
+
+    if (typeof payload.nbf !== "number") {
+      throw new ParseRequestError(
+        "Invalid client_assertion JWT: Invalid nbf claim in client JWT assertion"
+      );
+    }
+  }
+
+  if (!payload.iat) {
+    logger.warn(
+      "iat claim not included in client JWT assertion, this claim is recommended: https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/authenticate-your-user/#create-a-jwt"
+    );
+  }
+
+  if (payload.iat && typeof payload.iat !== "number") {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Invalid iat claim in client JWT assertion"
+    );
+  }
+
+  if (!payload.jti) {
+    logger.warn(
+      "No jti claim in client JWT assertion, this is required: https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/authenticate-your-user/#create-a-jwt"
+    );
+  }
+
+  if (tokenRequestClientId && tokenRequestClientId !== payload.sub) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: The client identifier doesn't match the client assertion subject"
+    );
+  }
+
+  if (!header.alg || !["RS256", "ES256"].includes(header.alg)) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: The client assertion JWT must be RSA or ECDSA-signed"
+    );
+  }
```

Comment:
I think these are the wrong way around? And the signed error message is `The client assertion JWT must be RSA or ECDSA-signed (RS256, RS384, RS512, PS256, PS384, PS512, ES256, ES384 or ES512)` unless I'm looking at the wrong one?

Reply:
Have replaced the error message with the one above and swapped these checks as they were in the wrong order